aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -1321,4 +1322,157 @@ Tensor any_sparse(const Tensor& self) {
   return at::any(self._values());
 }
 
+Tensor bmm_sparse_cpu(const SparseTensor& self, const Tensor& mat2) {
+  Tensor result = at::empty({}, mat2.options());
+  return bmm_out_sparse_cpu(result, self, mat2);
+}
+
+// Search a sorted strided array for the rightmost instance of a value.
+// Array must be sorted from lowest to highest
+template<typename scalar_t>
+scalar_t binary_search_strided_rightmost(scalar_t search_val, scalar_t* sorted_arr, int64_t sorted_arr_stride, int64_t length, bool* found) {
+  if (length == 0) {
+    *found = false;
+    return -1;
+  }
+
+  int64_t left_ind = 0;
+  int64_t right_ind = length - 1;
+  int64_t mid_ind;
+  bool done_searching = false;
+
+  while (!done_searching) {
+    mid_ind = (left_ind+right_ind) >> 1;
+    scalar_t mid_val = sorted_arr[mid_ind*sorted_arr_stride];
+
+    if (mid_val > search_val) {
+      right_ind = mid_ind-1;
+    } else if((mid_val == search_val) && (
+      (mid_ind == length - 1) || (sorted_arr[(mid_ind+1)*sorted_arr_stride] != search_val)
+    )) {
+      done_searching = true;
+      *found = true;
+    } else {
+      left_ind = mid_ind+1;
+    }
+
+    if (left_ind > right_ind) {
+      done_searching = true;
+      *found = false;
+      mid_ind = -1;
+    }
+  }
+
+  return mid_ind;
+}
+
+Tensor& bmm_out_sparse_cpu(Tensor& result, const SparseTensor& self, const Tensor& mat2) {
+  TORCH_CHECK(!mat2.is_sparse(), "bmm_sparse: Tensor 'mat2' must be dense");
+
+  TORCH_CHECK(self.dense_dim() == 0, "bmm_sparse: Tensor 'self' must have 0 dense dims, but has ", self.dense_dim());
```

Comment:
> Regarding how to define a batch of sparse matrices, we could accept a 3-D sparse tensor, just like how we're accepting a 3-D dense tensor. If we have a set of N 2-D tensors of identical shapes, the caller could torch.stack() them all into one sparse tensor.

I think this is probably right from an ease of API use perspective. When I think of blas-y APIs, they often let you pass in transpositions directly in the API, and that would have let you express a situation where you do pass in one dense dim at the end (which is the only thing sparse supports). I wonder if anyone will ask for this. (No action needed on this comment.)